packages/web_ui/src/components/SlavesPage.jsx

```diff
@@ -1,15 +1,27 @@
-import React, { useContext } from "react";
+import React, {useContext} from "react";
 
 import libLink from "@clusterio/lib/link";
 
 import DataTable from "./data-table";
 import ControlContext from "./ControlContext";
 import PageLayout from "./PageLayout";
 
-
 export default function SlavesPage() {
 	let control = useContext(ControlContext);
-
+	let table;
+	control.onLiveSlaveAdded("slaves_page", async (message) => {
+		let item = message.data.item;
+		let row = {
+			key: item["id"],
+			"Name": item["name"],
+			"Agent": item["agent"],
+			"Version": item["version"],
+			"Connected": item["connected"] && "Yes",
+		};
+		if (table) {
+			table.addRow(row);
+		}
+	});
```

Comment:
Modifying state directly inside a component function is a strict no-no in React. The rules for component functions is that the body may not have any side effects, here you modify the control instance in the body of the SlavesPage component. Side effects must be put inside a `useEffect` hook.

This also leaks the handler updating the table (it's not removed when the component is unmounted), and does not permit multiple instances of the component to exist at the same time as the "slaves_page" identifier is a singleton, which is bad.